client/material/tooltip.tsx

```diff
@@ -0,0 +1,220 @@
+  /** The text that will be used in the `aria-label` attribute. Defaults to the Tooltip's text. */
+  ariaLabel?: string
+  /** Class name applied to the component that wraps the Tooltip children. */
+  className?: string
+  /**
+   * A custom component that will be used instead of the default Tooltip content element. Can be
+   * used if you wish to customize the Tooltip style. Will get injected with the following props:
+   *  - $position: TooltipPosition
+   *  - children: string (the Tooltip's text)
+   * */
+  contentComponent?: React.ReactNode
+}
+
+/**
+ * A component that displays some content in a floated UI element when a user hovers over a target
+ * element. Should generally be used over native `title` attribute on elements.
+ *
+ * Utilizes popovers to deal with positioning, but with a much simpler API than that of a Popover.
+ */
+export function Tooltip({
+  text,
+  children,
+  className,
+  ariaLabel = text,
+  position = 'bottom',
+  contentComponent = <TooltipContent />,
+}: TooltipProps) {
+  const [open, setOpen] = useState(false)
+  const [anchorElem, setAnchorElem] = useState<HTMLElement>()
+  const transition = useTransition<boolean, UseTransitionProps<boolean>>(open, {
+    from: { opacity: 0, scale: 0.667 },
+    enter: { opacity: 1, scale: 1 },
+    leave: { opacity: 0, scale: 0.333 },
+    config: (item, index, phase) => key =>
+      phase === 'leave' || key === 'opacity' ? { ...defaultSpring, clamp: true } : defaultSpring,
+  })
+
+  const anchorOriginX = position === 'top' || position === 'bottom' ? 'center' : position
+  const anchorOriginY = position === 'left' || position === 'right' ? 'center' : position
+  let [, anchorX = 0, anchorY = 0] = useAnchorPosition(
+    anchorOriginX,
+    anchorOriginY,
+    anchorElem ?? null,
+  )
+
+  const onMouseEnter = useCallback((event: React.MouseEvent) => {
+    setAnchorElem(event.currentTarget as HTMLElement)
+  }, [])
+  const onMouseLeave = useCallback((event: React.MouseEvent) => {
+    setAnchorElem(undefined)
+  }, [])
+
+  useEffect(() => {
+    if (anchorElem) {
+      let timeout: ReturnType<typeof setTimeout> | undefined = setTimeout(() => {
+        timeout = undefined
+        setOpen(true)
+      }, 200)
+
+      return () => {
+        if (timeout) {
+          clearTimeout(timeout)
+        }
+      }
+    } else {
+      setOpen(false)
+      return () => {}
+    }
+  }, [anchorElem])
+
+  let originX: OriginX
+  if (anchorOriginX === 'left') {
+    originX = 'right'
+    anchorX = anchorX - 8
+  } else if (anchorOriginX === 'right') {
+    originX = 'left'
+    anchorX = anchorX + 8
+  } else {
+    originX = 'center'
+  }
+
+  let originY: OriginY
+  if (anchorOriginY === 'top') {
+    originY = 'bottom'
+    anchorY = anchorY - 8
+  } else if (anchorOriginY === 'bottom') {
+    originY = 'top'
+    anchorY = anchorY + 8
+  } else {
+    originY = 'center'
+  }
+
+  const tooltipContent = React.cloneElement(contentComponent as JSX.Element, {
```

Comment:
Nah, this was not the intention. The goal is that someone passes in a *component class/function*, not a whole element. You would then just use it as `<contentComponent $position={position}>{text}</contentComponent>` in the code below. In TS this will be a `React.ComponentType<{ $position, children }>`.

In general if you're using `React.cloneElement` I think you're probably doing the wrong thing, just for future reference. Extremely hard to ever manage types around those safely and they're just a pretty weird design

Reply:
Apparently TypeScript complains if a component doesn't start with a capitalized letter so I named the prop `ContentComponent`.